src/dbnode/storage/shard.go

```diff
@@ -2764,54 +2732,11 @@ func (s *dbShard) AggregateTiles(
 		return 0, err
 	}
 
-	var (
-		annotationPayload annotation.Payload
-		// NB: there is a maximum of 4 datapoints per frame for counters.
-		downsampledValues  = make([]downsample.Value, 0, 4)
-		processedTileCount int64
-		segmentCapacity    int
-		writerData         = make([][]byte, 2)
-		multiErr           xerrors.MultiError
-	)
-
-	for readerIter.Next() {
-		seriesIter, id, encodedTags := readerIter.Current()
-
-		seriesTileCount, err := encodeAggregatedSeries(seriesIter, annotationPayload, downsampledValues, encoder)
-		if err != nil {
-			s.metrics.largeTilesWriteErrors.Inc(1)
-			multiErr = multiErr.Add(err)
-			break
-		}
-
-		if seriesTileCount == 0 {
-			break
-		}
-
-		processedTileCount += seriesTileCount
-		segment := encoder.DiscardReset(opts.Start, segmentCapacity, targetSchemaDescr)
-
-		segmentLen := segment.Len()
-		if segmentLen > segmentCapacity {
-			// Will use the same capacity for the next series.
-			segmentCapacity = segmentLen
-		}
-
-		writerData[0] = segment.Head.Bytes()
-		writerData[1] = segment.Tail.Bytes()
-		checksum := segment.CalculateChecksum()
-
-		if err := writer.WriteAll(id, encodedTags, writerData, checksum); err != nil {
-			s.metrics.largeTilesWriteErrors.Inc(1)
-			multiErr = multiErr.Add(err)
-		} else {
-			s.metrics.largeTilesWrites.Inc(1)
-		}
-
-		segment.Finalize()
-	}
+	var multiErr xerrors.MultiError
 
-	if err := readerIter.Err(); err != nil {
+	processedTileCount, err := s.tileAggregator.AggregateTiles(
+		opts, targetNs, s.ID(), openBlockReaders, writer)
+	if err != nil {
```

Comment:
nit: add a comment here that we still need to do cleanup etc and should log any errors from that

Reply:
👍